exchange/exchange_test.go

```diff
@@ -1196,6 +1201,131 @@ func TestCategoryDedupe(t *testing.T) {
+
+	tempReqExt.Prebid.Targeting.DurationRangeSec = []int{15, 30, 50}
+	tempReqExt.Prebid.Targeting.IncludeBrandCategory.PrimaryAdServer = 2
+	tempReqExt.Prebid.Targeting.IncludeBrandCategory.Publisher = "some_publisher"
+
+	cats = []string{"IAB1-1"}
+	bid = openrtb.Bid{ID: "bid_id1", ImpID: "imp_id1", Price: 10.0000, Cat: cats, W: 1, H: 1}
+	bid1_1 = pbsOrtbBid{&bid, "video", nil, &openrtb_ext.ExtBidPrebidVideo{Duration: 30}}
+
+	seatBid = pbsOrtbSeatBid{innerBids, "USD", nil, nil}
+
+	bidCategory, adapterBids, rejections, err = applyCategoryMapping(nil, tempReqExt, adapterBids, categoriesFetcher, targData)
+
+	assert.Equal(t, nil, err, "Category mapping error should be empty")
+	assert.Equal(t, 1, len(rejections), "There should be 1 bid rejection message")
+	assert.Equal(t, "bid rejected [bid ID: bid_id1] reason: Category mapping file for primary ad server: 'dfp', publisher: 'some_publisher' not found", rejections[0], "Error message did not match expected result")
+	assert.Equal(t, 0, len(adapterBids[bidderName].bids), "Bidders number doesn't match")
+	assert.Equal(t, 0, len(bidCategory), "Bidders category mapping doesn't match")
+
+	cats = []string{"IAB1-1"}
+	bid = openrtb.Bid{ID: "bid_id1", ImpID: "imp_id1", Price: 10.0000, Cat: cats, W: 1, H: 1}
+	bid1_1 = pbsOrtbBid{&bid, "video", nil, &openrtb_ext.ExtBidPrebidVideo{Duration: 70}}
+
+	seatBid = pbsOrtbSeatBid{innerBids, "USD", nil, nil}
+
+	bidCategory, adapterBids, rejections, err = applyCategoryMapping(nil, requestExt, adapterBids, categoriesFetcher, targData)
+
+	assert.Equal(t, nil, err, "Category mapping error should be empty")
+	assert.Equal(t, 1, len(rejections), "There should be 1 bid rejection message")
+	assert.Equal(t, "bid rejected [bid ID: bid_id1] reason: Bid duration exceeds maximum allowed", rejections[0], "Error message did not match expected result")
+	assert.Equal(t, 0, len(adapterBids[bidderName].bids), "Bidders number doesn't match")
+	assert.Equal(t, 0, len(bidCategory), "Bidders category mapping doesn't match")
+
+	cats1 := []string{"IAB1-3"}
+	cats2 := []string{"IAB1-3"}
+	cats3 := []string{"IAB1-4"}
+	bid1 := openrtb.Bid{ID: "bid_id1", ImpID: "imp_id1", Price: 10.0000, Cat: cats1, W: 1, H: 1}
+	bid2 := openrtb.Bid{ID: "bid_id2", ImpID: "imp_id2", Price: 10.0000, Cat: cats2, W: 1, H: 1}
+	bid3 := openrtb.Bid{ID: "bid_id3", ImpID: "imp_id3", Price: 10.0000, Cat: cats3, W: 1, H: 1}
+
+	bid1_1 = pbsOrtbBid{&bid1, "video", nil, &openrtb_ext.ExtBidPrebidVideo{Duration: 30}}
+	bid1_2 := pbsOrtbBid{&bid2, "video", nil, &openrtb_ext.ExtBidPrebidVideo{Duration: 30}}
+	bid1_3 := pbsOrtbBid{&bid3, "video", nil, &openrtb_ext.ExtBidPrebidVideo{Duration: 30}}
+
+	innerBids = []*pbsOrtbBid{
+		&bid1_1,
+		&bid1_2,
+		&bid1_3,
+	}
+
+	seatBid = pbsOrtbSeatBid{innerBids, "USD", nil, nil}
+
+	bidCategory, adapterBids, rejections, err = applyCategoryMapping(nil, requestExt, adapterBids, categoriesFetcher, targData)
+
+	// There is a random factor to deduplication
```

Comment:
Random and Tests aren't a good match. Think of what is the cause of the randomness and how it can be factored away for the test. If it's just order dependent, consider using assert statements which ignore order.. like `assert.Contains`.

Reply:
Maybe passing in a function that decides which bid to keep instead of hard-coding it to a 50% chance there where it picks (so test could use a function that just always picks the first)? I kind of wanted to re-work that code eventually anyway, not a fan of how the original bid would have to win many coin flips in a row to remain there if there were like 4 or 5 duplicates found (probably unlikely, but it's still one of those things that bugs me to this day).

Reply:
Updated this test here to instead use the same bid ID so that the expected messages are always the same. The removal of bids uses the index instead of the ID, so I am just using two bids that are the same and expecting 1 to be removed, and 1 to be in the resulting bidCategory. Let me know if I should use a different approach (also could use the `assert.Regex` if we want it to have different IDs).

Reply:
Same bid id = same expected message, that's a super simple way to work around the randomness with duplicate bid removal, I think it's a great idea!